backend/src/schema/resolvers/reports.js

```diff
@@ -1,62 +1,76 @@
       const session = driver.session()
-      const reportProperties = {
-        id: reportId,
-        createdAt: new Date().toISOString(),
-        reasonCategory,
-        reasonDescription,
-      }
-
-      const reportQueryRes = await session.run(
-        `
-          MATCH (u:User {id:$submitterId})-[:REPORTED]->(report)-[:REPORTED]->(resource {id: $resourceId}) 
-          RETURN labels(resource)[0] as label
-        `,
-        {
-          resourceId,
-          submitterId: user.id,
-        },
-      )
-      const [rep] = reportQueryRes.records.map(record => {
-        return {
-          label: record.get('label'),
-        }
+      const writeTxResultPromise = session.writeTransaction(async txc => {
+        const reportRelationshipTransactionResponse = await txc.run(
+          `
+            MATCH (submitter:User {id: $submitterId})
+            MATCH (resource {id: $resourceId})
+            WHERE resource:User OR resource:Comment OR resource:Post
+            CREATE (resource)<-[report:REPORTED {createdAt: $createdAt, reasonCategory: $reasonCategory, reasonDescription: $reasonDescription}]-(submitter)
+            RETURN report, submitter, resource, labels(resource)[0] as type
+          `,
+          {
+            resourceId,
+            submitterId: user.id,
+            createdAt: new Date().toISOString(),
+            reasonCategory,
+            reasonDescription,
+          },
+        )
+        return reportRelationshipTransactionResponse.records.map(record => ({
+          report: record.get('report'),
+          submitter: record.get('submitter'),
+          resource: record.get('resource').properties,
+          type: record.get('type'),
+        }))
       })
-
-      if (rep) {
-        throw new Error(rep.label)
+      try {
+        const txResult = await writeTxResultPromise
+        if (!txResult[0]) return null
+        const { report, submitter, resource, type } = txResult[0]
+        createdRelationshipWithNestedAttributes = {
+          ...report.properties,
+          post: null,
+          comment: null,
+          user: null,
+          submitter: submitter.properties,
+          type,
+        }
+        switch (type) {
+          case 'Post':
+            createdRelationshipWithNestedAttributes.post = resource
+            break
+          case 'Comment':
+            createdRelationshipWithNestedAttributes.comment = resource
+            break
+          case 'User':
+            createdRelationshipWithNestedAttributes.user = resource
+            break
+        }
+      } finally {
+        session.close()
       }
-
+      return createdRelationshipWithNestedAttributes
+    },
+  },
+  Query: {
+    reports: async (_parent, _params, { driver }, _resolveInfo) => {
+      const session = driver.session()
       const res = await session.run(
         `
-          MATCH (submitter:User {id: $userId})
-          MATCH (resource {id: $resourceId})
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T09:35:13Z" title="Monday, October 14th 2019, 11:35:13 am +02:00">Oct 14, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -87,6 +81,60 @@ export default {
           break
       }
 
+      return response
+    },
+  },
+  Query: {
+    reports: async (_parent, _params, { driver, _req, _user }, _resolveInfo) => {
+      const session = driver.session()
+      const res = await session.run(
+        `
+          MATCH (submitter:User)-[report:REPORTED]->(resource)
+          WHERE resource:User OR resource:Comment OR resource:Post
+          RETURN report, submitter, resource, labels(resource)[0] as type
+        `,
+        {},
+      )
+      session.close()
+
+      const dbResponse = res.records.map(r => {
+        return {
+          report: r.get('report'),
+          submitter: r.get('submitter'),
+          resource: r.get('resource'),
+          type: r.get('type'),
+        }
+      })
+      if (!dbResponse) return null
+
+      const response = []
+      dbResponse.forEach(ele => {
+        const { report, submitter, resource, type } = ele
+
+        const responseEle = {
+          ...report.properties,
+          resource: resource.properties,
+          resourceId: resource.properties.id,
```


why do we need `resource` and also `resourceId` if the resource has the id already?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-14T09:36:40Z" title="Monday, October 14th 2019, 11:36:40 am +02:00">Oct 14, 2019</time>_

---


why do we need `resource` at all actually, aren't we checking the type and returning the resources' properties based on which type it is?

Reply:
> [<img alt="Tirokk" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/Tirokk) **Authored by [Tirokk](https://github.com/Tirokk)**
_<time datetime="2019-10-14T13:40:26Z" title="Monday, October 14th 2019, 3:40:26 pm +02:00">Oct 14, 2019</time>_

---


I forgot to delete the code here after I find out it can not be supported now … Thanks for the hint @mattwr18 
Done …